int32 location_state
int32 angle_state
